standards/practices/testing.md

```diff
@@ -0,0 +1,198 @@
+    }
+}
+```
+
+**Bad**
+```php
+<?php
+
+namespace Tests\Unit;
+
+use App\ClientUser;
+use App\User;
+use Facades\App\Library\Helpers\AuthHelper;
+use Illuminate\Foundation\Testing\RefreshDatabase;
+use Tests\TestCase;
+
+class AuthHelperTest extends TestCase
+{
+    use RefreshDatabase;
+    
+    function testUsersReturnsAdminUsersButClientUsersReturnsClientUsers()
+    {    
+        $this->assertEquals('client-users', AuthHelper::getAuthProviderKey(factory(User::class)->create()->email));
+        $this->assertEquals('admin-users', AuthHelper::getAuthProviderKey(factory(ClientUser::class)->create()->email));
+    }
+}
+```
+
+This test is bad for a number of reasons:
+
+1. It's supposed to be a [unit test](#unit-tests), but it depends on outside systems (the database and User model).
+2. One test covers two different actions.
+3. The test name is difficult to read in camelCase.
+4. The test mixes all stages together.
+5. The code covered is so simple it's not worth testing (little more than third-party code hidden behind a simple ternary).
+
+**Good**
+```php
+<?php
+
+namespace Tests\Feature;
+
+use App\ClientUser;
+use App\User;
+use Illuminate\Foundation\Testing\RefreshDatabase;
+use Tests\TestCase;
+
+class AuthenticationTest extends TestCase
+{
+    use RefreshDatabase;
+
+    /** @test */
+    public function admins_can_log_in()
+    {
+        // Arrange
+        // Create an admin.
+        $user = factory(User::class)->create(['password' => bcrypt('pw1234')]);
+
+        // Act
+        // Try to log in.
+        $response = $this->post('/login', [
+            'email'    => $user->email,
+            'password' => 'pw1234'
+        ]);
+
+        // Assert
+        // Ensure user logged in.
+        $response->assertRedirect(route('home'));
+        $this->assertAuthenticatedAs($user);
+    }
+
+    /** @test */
+    public function client_users_can_log_in()
+    {
+        // Arrange
+        // Create a client user.
+        $user = factory(ClientUser::class)->create(['password' => bcrypt('pw1234')]);
+
+        // Act
+        // Try to log in.
+        $response = $this->post('/login', [
+            'email'    => $user->email,
+            'password' => 'pw1234'
+        ]);
+
+        // Assert
+        // Ensure user logged in.
+        $response->assertRedirect(route('home'));
+        $this->assertAuthenticatedAs($user);
+    }
+}
+```
+
+*In most frameworks, it's not worth testing authentication, because the framework handles pretty much everything. However in this example, the authentication is customized enough that we need to avoid regressions.*
+
+These two tests improve on the previous example in several ways:
+
+1. We converted from a unit test to an integration test (called a feature test in Laravel). The feature depends on several systems more simply tested with a couple short integration tests than with several unit tests that can't reliably test the whole feature.
```

Comment:
Should items 1-5 be converted into set of recommendations/suggestions or should it be a case of 'learning from example'? If so, should there be more examples?

Reply:
I see these as recommendations through an example. I'm not entirely opposed to it explicitly saying, "These are recommendations." However, I think it's reasonably clear that when presenting reasons why something is good, those reasons are things to strive for.

I don't think there _should_ be more examples, but I do think there _could_ be more examples. I think if we want to add more examples, we should do that outside of this proposal as that would unnecessarily slow this process down.

I'd likely also want to maintain any further examples in (a) separate file(s), b/c [the standards are supposed to be short](https://github.com/HighlandSolutions/development-standards/tree/master/standards#we-prefer-fewer-simpler-standards). I hope this is the longest standard we end up with. The only reason I'm comfortable with this length is that testing is so important and complex.